quickstart/src/configurations/aks/clusters.tf

```diff
@@ -2,4 +2,7 @@ module "aks_zero" {
   source = "github.com/kbst/terraform-kubestack//azurerm/cluster?ref={{version}}"
 
   configuration = var.clusters["aks_zero"]
+
+  # vnet_subnet_id = azurerm_subnet.external.id  # uncomment and populate with an externally-created
```

Comment:
Up to now, each `cluster` module has only the [three common module attributes](https://www.kubestack.com/framework/documentation/cluster-pairs#module-attributes).

All additional configuration is wrapped inside the configuration attribute, which allows the inheritance between the environments.

By adding it here, you can reference the ID from another TF resource, but lose the ability to leverage inheritance to have different IDs per environment.

I generally err on the side of keeping the API of the provider specific cluster modules the same.

What options do we have?

Reply:
Defining the vnet inside the module seems like it will limit users too much. So configuring it outside and passing the ID in seems the best approach based on what we know so far and is what you implemented so far. Maybe a good compromise is to simply move the attribute inside the configuration with the other ones for now and then document that it is important to overwrite this per environment.

This is a limitation of the current way how the inheritance is implemented. I think there may be better ways, this was built pre 0.12. But I haven't looked into it and it would certainly be a bigger change so it should be tackled in its own PR.

What do you think about moving it into the configuration for now. Then I can start testing this and if everything is fine I can merge it.

Reply:
When you suggest moving the attribute inside the configuration, what attribute are you referring to specifically? If you mean the vnet_subnet_id, it's referring to a value that is not known until the resources are created. I get the following when I try to point to that ID within config.auto.tfvars:

![image](https://user-images.githubusercontent.com/317047/94845769-c4dacd80-03e5-11eb-95fe-28e0ebcf2071.png)

Reply:
Yes, the value would need to be the hardcoded string of the id inside the tfvars file. It can't be a reference, that's what the error says.

Reply:
So the vision I've been pursuing is one where I can maintain my infrastructure entirely with terraform, which includes PaaS data services like CosmosDB or MSSQL, as well as Virtual Network Gateways to provide access to on-premises resources. For these resources external to Kubestack, my intention was to maintain the apps/ops/loc workspaces, and ensure that the external resources are created so that they are isolated appropriately in the various environments. Putting the subnet ID in config means that at best, the external resources have to live in a separate terraform module/project, and cannot be deployed in a single terraform apply with the rest of the kubestack resources.

Given this, I would be more interested in moving the custom vnet into the module and exposing it via outputs to allow users to associate it to other PaaS resources, if needed/desired.

Reply:
I am fine with that too.

Reply:
I have made this update to the PR. Configuration variables now contain a set of variables that are used to create a vnet and subnet, and to pass them into the azurerm_kubernetes_cluster resource appropriately.

I also engaged with Microsoft support with a question on the behavior of the NSG, and learned that when using Advanced Networking in a cluster, the NSG gets associated with the VM Scale Set's NICs, rather than the subnet generally. This meant that we can safely remove the work-around script that was present in previous iterations of this PR.

Reply:
Sounds great, I'll take a look tomorrow.